query_executor.go

```diff
@@ -50,51 +70,148 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
 			continue
 		}
 
-		iter = q.attemptQuery(qry, conn)
-		// Update host
-		hostResponse.Mark(iter.err)
+		// store the result
+		q.specWG.Add(1)
+		go q.runAndStore(qry, conn, selectedHost, results, nil)
+
+		// fetch from the channel whatever is there
+		res = <-results
 
-		if rt == nil {
-			iter.host = host
-			break
+		if res.retryType == RetryNextHost {
+			continue
 		}
 
-		switch rt.GetRetryType(iter.err) {
-		case Retry:
-			for rt.Attempt(qry) {
-				iter = q.attemptQuery(qry, conn)
-				hostResponse.Mark(iter.err)
-				if iter.err == nil {
-					iter.host = host
-					return iter, nil
-				}
-				if rt.GetRetryType(iter.err) != Retry {
-					break
-				}
-			}
-		case Rethrow:
-			return nil, iter.err
-		case Ignore:
-			return iter, nil
-		case RetryNextHost:
-		default:
+		return res.iter, res.err
+
+	}
+
+	if res.iter == nil {
+		return nil, ErrNoConnections
+	}
+
+	return res.iter, nil
+}
+
+func (q *queryExecutor) executeSpeculativeQuery(qry ExecutableQuery) (*Iter, error) {
+	hostIter := q.policy.Pick(qry)
+	sp := qry.speculativeExecutionPolicy()
+
+	results := make(chan queryResponse)
+
+	stopExecutions := make(chan struct{})
+	defer close(stopExecutions)
+
+	var (
+		res             queryResponse
+		specExecCounter int
+	)
+	for selectedHost := hostIter(); selectedHost != nil && specExecCounter < sp.Executions(); selectedHost = hostIter() {
```

Comment:
I'm not sure I like this logic very much. I was thinking this would look something more like

```go
timer := time.NewTimer(speculation.After())
defer timer.Stop()
result := make(chan *Iter, speculation.Attempts())
go queryAttempt(ctx, hostIter, result)
for i := 0; i < speculation.Attempts(); i++ {
    select {
    case <-timer.C:
         go queryAttempt(ctx, hostIter, result)
         timer.Result(speculation.After())
    case result := <-result:
        return result
    case ctx.Done():
        return ctx.Err()
    }
}
```

And have all retrying logic handled within `queryAttempt` somehow.

This may require adjustments to the hostIter interface to say if we have `more` hosts to try

Reply:
Should query speculation use the same host iterator for each query or a new one?

Reply:
Considering that each speculative "execution" is defined per node, I think we need the same host iterator in order to go only once over all the endpoints. We don't want a different query to try a node that has already been tried.